### PR TITLE
Issue / Multi Authentication

### PR DESCRIPTION
**Pull request by dncsvr**
Learn how to authenticate requests using multiple authentication schemes

## Tasks

- [x] Add a sample project for multi authentication
  - [x] multiple schemes to authenticate a request
    - [x] using `Authorize` attribute schemes
    - [x] using authorization policy with multiple schemes
  - [x] using multiple handlers within a scheme
  - [x] review samples and validate learn scope
  - [x] documentation

### CODE REVIEW

File: authentication/MultiAuthentication/Handlers/OrganizationIdAuthenticationHandler.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+public class OrganizationIdAuthenticationHandler(
+    IOptionsMonitor<AuthenticationSchemeOptions> options,
+    ILoggerFactory logger,
+    UrlEncoder encoder
+) : AuthenticationHandler<AuthenticationSchemeOptions>(options, logger, encoder)
+{
+    protected override Task<AuthenticateResult> HandleAuthenticateAsync()
+    {
+        if (Context.Request.Headers.TryGetValue("X-ORGANIZATION-ID", out StringValues key))
```
Comment (cihandeniz) [inline authentication/MultiAuthentication/Handlers/OrganizationIdAuthenticationHandler.cs:17]: http headers should be in `X-Header-Name` case

File: authentication/MultiAuthentication/MultiAuthenticationExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            manager.ApplicationParts.Add(new SingleControllerApplicationPart<SchemeSelectorController>());
+        });
+
+        source.AddAuthentication(options =>
+        {
+            options.DefaultScheme = "ApiKey";
+            options.DefaultAuthenticateScheme = "ApiKey";
+            options.AddScheme<ApiKeyAuthenticationHandler>("ApiKey", "ApiKey");
+            options.AddScheme<BearerTokenAuthenticationHandler>("BearerToken", "BearerToken");
+            options.AddScheme<OrganizationIdAuthenticationHandler>("OrganizationId", "OrganizationId");
```
Comment (cihandeniz) [inline authentication/MultiAuthentication/MultiAuthenticationExtensions.cs:28]: rename these schemes to have abstract names `X`, `Y`, `Z`

File: authentication/MultiAuthentication/Handlers/BearerTokenAuthenticationHandler.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    UrlEncoder encoder
+) : AuthenticationHandler<AuthenticationSchemeOptions>(options, logger, encoder)
+{
+    protected override Task<AuthenticateResult> HandleAuthenticateAsync()
+    {
+        if (Context.Request.Headers.Authorization.Any())
+        {
+            var givenToken = $"{Context.Request.Headers.Authorization}".Replace("Bearer", string.Empty).Trim();
+
+            var claim = new Claim("Token", $"{givenToken}");
```
Comment (cihandeniz) [inline authentication/MultiAuthentication/Handlers/BearerTokenAuthenticationHandler.cs:20]: Token is not a claim, don't add it to user

File: authentication/.editorconfig [hunk 4]
```diff
@@ -0,0 +1,308 @@
+root = true
```
Comment (cihandeniz) [inline authentication/.editorconfig:1]: move this file to root, near `.sln` file

File: authentication/MultiAuthentication/Controllers/PolicyAndSchemesController.cs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    [Route("multi-authentication-scheme")]
+    public object MultiAuthenticationScheme()
+    {
+        var principal = httpContextAccessor.HttpContext?.User ?? throw new();
+
+        return principal.ToDictionary();
+    }
+
+    [HttpGet]
+    [Authorize(Policy = "Backend")]
```
Comment (cihandeniz) [inline authentication/MultiAuthentication/Controllers/PolicyAndSchemesController.cs:44]: we need to have a test case where it uses a specific scheme while requiring a specific policy

File: authentication/MultiAuthentication/MultiAuthenticationExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.ApplicationParts;
+using MultiAuthentication.Controllers;
+using MultiAuthentication.Handlers;
+using System.Reflection;
+using System.Security.Claims;
+
+namespace MultiAuthentication;
+
+public static class MultiAuthenticationExtensions
```
Comment (cihandeniz) [inline authentication/MultiAuthentication/MultiAuthenticationExtensions.cs:12]: let's leave only the option that we decided to use, and remove the other from code.

it's ok to mention the other option in document as long as it is very very brief

File: authentication/README.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
+# Authentication
+
+## Multi Authentication
```
Comment (cihandeniz) [inline authentication/README.md:3]: need to add a link to this in the root `README.md`

File: authentication/README.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+First option is to set _AuthenticationSchemes_ value of `Authorize` attribute
+and register multiple schemes with `AuthenticationBuilder`.
+
+```csharp
+//Program.cs
+
+builder.Services.
+    .AddAuthentication()
+    .AddScheme<AuthenticationSchemeOptions, HandlerOne>("SchemeOne", options => { })
+    .AddScheme<AuthenticationSchemeOptions, HandlerTwo>("SchemeTwo", options => { });
```
Comment (cihandeniz) [inline authentication/README.md:21]: this is the abstract naming i'm talking about. and `SchemeOne` is even better than what i've suggested `X`, `Y`

File: authentication/README.md [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+[Authorize(AuthenticationSchemes = "ShemeOne,SchemeTwo")]
+public void Action()
+{
+    ...
+}
+```
+
+> [!NOTE]
+>
+> The a request is made to the endpoint, will be handled by all two named schemes
```
Comment (cihandeniz) [inline authentication/README.md:37]: The a request :thinking:

File: authentication/MultiAuthentication/MultiAuthenticationExtensions.cs (hunk 2 above)
Comment (dncsvr) [reply authentication/MultiAuthentication/MultiAuthenticationExtensions.cs]: renamed the schemes to "Default", "Alternative", and "Anonymous"

File: authentication/README.md (hunk 8 above)
Comment (dncsvr) [reply authentication/README.md]: changed adding schemes to reflect sample code using "Default","Alternative" and "Anonymous" names

**Review by cihandeniz — APPROVED** [review]

**Pull request merged**